### PR TITLE
Add restart deployment and spawn job actions

### PR DESCRIPTION
**Pull request by ashu8912**

### CODE REVIEW

**Review by joaquimrocha — CHANGES_REQUESTED** [review]

File: frontend/src/components/common/Resource/RestartButton.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        cancelledMessage: t('Cancelled restarting {{ itemName }}.', { itemName }),
+        successMessage: t('Restarted {{ itemName }}.', { itemName }),
+        errorMessage: t('Failed to restart {{ itemName }}.', { itemName }),
+        cancelUrl,
+        errorUrl: cancelUrl,
+      })
+    );
+  }
+
+  if (!['Deployment', 'StatefulSet', 'DaemonSet'].includes(item.kind)) {
```
[the remark this answers is not in the document]
Comment (ashu8912) [reply frontend/src/components/common/Resource/RestartButton.tsx:67]: updated

File: frontend/src/components/common/Resource/RestartButton.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    try {
+      const clonedItem = _.cloneDeep(item);
+      clonedItem.spec.template.metadata.annotations = {
+        ...clonedItem.metadata.annotations,
+        'kubectl.kubernetes.io/restartedAt': new Date().toISOString(),
+      };
+      apply(clonedItem.jsonData);
+    } catch (err) {
+      throw err;
+    }
```
[the remark this answers is not in the document]
Comment (ashu8912) [reply frontend/src/components/common/Resource/RestartButton.tsx]: updated with not throwing but logging if there is a error

File: frontend/src/components/common/Resource/RestartButton.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  if (!['Deployment', 'StatefulSet', 'DaemonSet'].includes(item.kind)) {
+    return null;
+  }
+
+  return (
+    <AuthVisible
+      item={item}
+      authVerb="update"
+      onError={(err: Error) => {
+        console.error(`Error while getting authorization for scaling button in ${item}:`, err);
```
[the remark this answers is not in the document]
Comment (ashu8912) [reply frontend/src/components/common/Resource/RestartButton.tsx]: updated

File: frontend/src/components/cronjob/Details.tsx [hunk 4]
```diff
@@ -22,10 +34,16 @@ export default function CronJobDetails() {
   const [cronJob, cronJobError] = CronJob.useGet(name, namespace);
   const [isCronSuspended, setIsCronSuspended] = useState(false);
   const [isCheckingCronSuspendStatus, setIsCheckingCronSuspendStatus] = useState(true);
+  const [openJobDialog, setOpenJobDialog] = useState(false);
   const dispatch = useDispatch();
 
   function applyFunc(newItem: KubeObjectInterface) {
-    setIsCheckingCronSuspendStatus(true);
+    console.log('resource is', newItem);
```
[the remark this answers is not in the document]
Comment (ashu8912) [reply frontend/src/components/cronjob/Details.tsx]: udpated

File: frontend/src/components/common/Resource/RestartButton.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import { clusterAction } from '../../../redux/actions/actions';
+import AuthVisible from './AuthVisible';
+
+interface RestartButtonProps {
+  item: KubeObject;
+}
+
+export function RestartButton(props: RestartButtonProps) {
+  const { item } = props;
+  const { t } = useTranslation();
```
Comment (joaquimrocha) [inline frontend/src/components/common/Resource/RestartButton.tsx:27]: This needs a namespace. 'frequent' I guess.

File: frontend/src/components/common/Resource/RestartButton.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+interface RestartDialogProps {
+  resource: KubeObject;
+  open: boolean;
+  onClose: () => void;
+  onSave: () => void;
+}
+
+function RestartDialog(props: RestartDialogProps) {
+  const { resource, open, onClose, onSave } = props;
+  const { t } = useTranslation();
```
Comment (joaquimrocha) [inline frontend/src/components/common/Resource/RestartButton.tsx:98]: This needs the namespace again. 'frequent'

File: frontend/src/components/cronjob/Details.tsx [hunk 7]
```diff
@@ -54,6 +71,103 @@ export default function CronJobDetails() {
...
+  const uniqueString = () => {
+    const timestamp = Date.now().toString(36);
+    const randomNum = Math.random().toString(36).substr(2, 5);
+    return `${timestamp}-${randomNum}`;
+  };
+
+  function SpawnJobDialog() {
+    const job = _.cloneDeep(cronJob.spec.jobTemplate);
+    const [jobName, setJobName] = useState(
+      `${cronJob?.metadata?.name}-${'manual-job-'}${uniqueString()}`
```
Comment (joaquimrocha) [inline frontend/src/components/cronjob/Details.tsx:84]: You don't need this `${'manual-job-'}` as a variable.
Nit: Maybe `-manual-spawn-` makes more sense.

File: frontend/src/components/cronjob/Details.tsx [hunk 8]
```diff
@@ -54,6 +71,103 @@ export default function CronJobDetails() {
...
   const ownedJobs = filterOwnedJobs(jobs);
 
+  // method to generate a unique string
+  const uniqueString = () => {
+    const timestamp = Date.now().toString(36);
+    const randomNum = Math.random().toString(36).substr(2, 5);
+    return `${timestamp}-${randomNum}`;
+  };
+
+  function SpawnJobDialog() {
```
Comment (joaquimrocha) [inline frontend/src/components/cronjob/Details.tsx:81]: I'd declare this dialog outside of the CronJobDetails if possible.

File: frontend/src/components/cronjob/Details.tsx [hunk 9]
```diff
@@ -22,10 +34,15 @@ export default function CronJobDetails() {
   const [cronJob, cronJobError] = CronJob.useGet(name, namespace);
   const [isCronSuspended, setIsCronSuspended] = useState(false);
   const [isCheckingCronSuspendStatus, setIsCheckingCronSuspendStatus] = useState(true);
+  const [openJobDialog, setOpenJobDialog] = useState(false);
   const dispatch = useDispatch();
 
   function applyFunc(newItem: KubeObjectInterface) {
-    setIsCheckingCronSuspendStatus(true);
+    if (newItem.kind === 'CronJob') {
+      setIsCheckingCronSuspendStatus(true);
```
Comment (joaquimrocha) [inline frontend/src/components/cronjob/Details.tsx:42]: Will this ever happen? I read here that you only run it with a Job kind.

File: frontend/src/components/cronjob/Details.tsx [hunk 10]
```diff
@@ -93,18 +207,22 @@ export default function CronJobDetails() {
 
   const actions = [];
 
-  actions.push(
+  actions.push([
     <AuthVisible authVerb="modify" item={cronJob}>
       <ActionButton
-        description={isCronSuspended ? t('frequent|Resume') : t('frequent|Suspend')}
+        description={t('frequent|Spawn Job')}
```
Comment (joaquimrocha) [inline frontend/src/components/cronjob/Details.tsx:213]: I think this `AuthVisible` component needs to check if we can create Job resources, not modify CronJob ones, right?

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Found a few things in the code but the functionality is working fine for me.
I noticed though that if the CronJob changes when the spawn dialog is showing (e.g. by waiting for it to keep creating jobs, or by clicking pause/resume and quickly pressing "spawn"), there's a lot of flickering, like if the spawn dialog keeps getting hidden and re-shown.

File: frontend/src/components/cronjob/Details.tsx (hunk 9 above)
Comment (ashu8912) [reply frontend/src/components/cronjob/Details.tsx:42]: This is run for changing the suspend state too for CronJob

File: frontend/src/components/cronjob/Details.tsx (hunk 10 above)
Comment (ashu8912) [reply frontend/src/components/cronjob/Details.tsx:225]: updated

File: frontend/src/components/cronjob/Details.tsx (hunk 8 above)
Comment (ashu8912) [reply frontend/src/components/cronjob/Details.tsx]: updated

File: frontend/src/components/cronjob/Details.tsx (hunk 7 above)
Comment (ashu8912) [reply frontend/src/components/cronjob/Details.tsx]: updated

File: frontend/src/components/common/Resource/RestartButton.tsx (hunk 6 above)
Comment (ashu8912) [reply frontend/src/components/common/Resource/RestartButton.tsx]: updated

File: frontend/src/components/common/Resource/RestartButton.tsx (hunk 5 above)
Comment (ashu8912) [reply frontend/src/components/common/Resource/RestartButton.tsx]: updated

File: frontend/src/components/cronjob/Details.tsx (hunk 10 above)
Comment (joaquimrocha) [reply frontend/src/components/cronjob/Details.tsx]: This wasn't really changed to check the Job resource but I will do it.

**Comment by ashu8912** [comment]
@joaquimrocha  tested with the new changes and the code also makes more sense now, thanks

**Pull request merged**